File: apps/mobile/lib/features/proof/data/proof_file_store.dart
// Habitat Proof File Storage Engine
import 'dart:convert';
import 'dart:io';
import 'package:crypto/crypto.dart';
import 'package:path/path.dart' as p;

class ProofFileStore {
  final Map<String, List<int>> _memoryStorage = {};
  final String? _baseDirectory;

  ProofFileStore({String? baseDirectory}) : _baseDirectory = baseDirectory;

  factory ProofFileStore.create({String? baseDirectory}) {
    return ProofFileStore(baseDirectory: baseDirectory);
  }

  String computeSha256(List<int> bytes) {
    return sha256.convert(bytes).toString();
  }

  Future<String> saveProofBytes({
    required String taskId,
    required String attemptId,
    required List<int> bytes,
    required String extension,
  }) async {
    final fileName = '${taskId}_$attemptId.$extension';
    final targetPath = _baseDirectory != null
        ? p.join(_baseDirectory!, fileName)
        : '/mock_proofs/$fileName';

    if (_baseDirectory != null) {
      final file = File(targetPath);
      await file.parent.create(recursive: true);
      await file.writeAsBytes(bytes);
    } else {
      _memoryStorage[targetPath] = bytes;
    }

    return targetPath;
  }

  Future<bool> proofExists(String path) async {
    if (_baseDirectory != null) {
      return File(path).exists();
    }
    return _memoryStorage.containsKey(path);
  }

  Future<int> getFileByteSize(String path) async {
    if (_baseDirectory != null) {
      final file = File(path);
      if (await file.exists()) {
        return await file.length();
      }
      return 0;
    }
    return _memoryStorage[path]?.length ?? 0;
  }

  Future<void> deleteProof(String path) async {
    if (_baseDirectory != null) {
      final file = File(path);
      if (await file.exists()) {
        await file.delete();
      }
    } else {
      _memoryStorage.remove(path);
    }
  }
}

File: backend/src/modules/proofs/domain/proof-file-store.ts
// Production Proof File Storage Engine
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';

export interface ProofFileMetadata {
  storageKey: string;
  filePath: string;
  fileSizeBytes: number;
  sha256: string;
  createdAt: string;
  mimeType: string;
}

export class ProofFileStore {
  private static baseStorageDir = path.resolve(process.cwd(), 'data', 'proofs');

  public static setStorageDirForTesting(customDir: string): void {
    this.baseStorageDir = customDir;
  }

  public static getStorageDir(): string {
    return this.baseStorageDir;
  }

  public static computeSha256(bytes: Buffer): string {
    return crypto.createHash('sha256').update(bytes).digest('hex');
  }

  public static async saveProof(params: {
    missionId: string;
    proofId: string;
    bytes: Buffer;
    extension?: string;
    mimeType?: string;
  }): Promise<ProofFileMetadata> {
    const ext = (params.extension || 'bin').replace(/^\./, '');
    const missionDir = path.join(this.baseStorageDir, params.missionId);

    if (!fs.existsSync(missionDir)) {
      fs.mkdirSync(missionDir, { recursive: true });
    }

    const filename = `${params.proofId}.${ext}`;
    const filePath = path.join(missionDir, filename);
    const storageKey = `proofs/${params.missionId}/${filename}`;

    fs.writeFileSync(filePath, params.bytes);

    const sha256 = this.computeSha256(params.bytes);

    return {
      storageKey,
      filePath,
      fileSizeBytes: params.bytes.length,
      sha256,
      createdAt: new Date().toISOString(),
      mimeType: params.mimeType || 'application/octet-stream'
    };
  }

  public static readProof(storageKey: string): Buffer {
    const filePath = this.resolvePath(storageKey);
    if (!fs.existsSync(filePath)) {
      throw new Error(`FILE_NOT_FOUND: Proof file not found at ${storageKey}`);
    }
    return fs.readFileSync(filePath);
  }

  public static exists(storageKey: string): boolean {
    const filePath = this.resolvePath(storageKey);
    return fs.existsSync(filePath);
  }

  public static deleteProof(storageKey: string): boolean {
    const filePath = this.resolvePath(storageKey);
    if (fs.existsSync(filePath)) {
      fs.unlinkSync(filePath);
      return true;
    }
    return false;
  }

  public static getMetadata(storageKey: string): ProofFileMetadata | null {
    const filePath = this.resolvePath(storageKey);
    if (!fs.existsSync(filePath)) return null;

    const stats = fs.statSync(filePath);
    const bytes = fs.readFileSync(filePath);
    const sha256 = this.computeSha256(bytes);

    return {
      storageKey,
      filePath,
      fileSizeBytes: stats.size,
      sha256,
      createdAt: stats.birthtime.toISOString(),
      mimeType: storageKey.endsWith('.mp4')
        ? 'video/mp4'
        : storageKey.endsWith('.jpg') || storageKey.endsWith('.jpeg')
          ? 'image/jpeg'
          : 'application/octet-stream'
    };
  }

  private static resolvePath(storageKey: string): string {
    const cleanKey = storageKey.replace(/^proofs\//, '');
    return path.join(this.baseStorageDir, cleanKey);
  }
}
